(SERVER-1630) Add support for JRuby 9k

camlow325 opened the pull request:
This PR adds support for being able to run Puppet Server with JRuby 9k.  The
PR only currently reworks dependencies in the project.clj file such that, via the
use of a jruby9k profile, it is possible to exercise the repl and tests from source
with JRuby 9k instead of the default, JRuby 1.7.  We won't be able to merge this
PR until additional work can be added onto it for enabling the ezbake process
to build packages which have configurable support for JRuby 1.7 and 9k with
Puppet Server, work being done for [SERVER-1772](https://tickets.puppetlabs.com/browse/SERVER-1772).

camlow325 commented on project.clj:
```diff
@@ -62,7 +62,11 @@
                  [net.logstash.logback/logstash-logback-encoder]
 
-                 [puppetlabs/jruby-utils "0.8.0"]
+                 [puppetlabs/jruby-utils "0.8.1-SNAPSHOT"
```
This PR will fail Travis for now b/c of a dependency on work being done in a [related jruby-utils PR](https://github.com/puppetlabs/jruby-utils/pull/45) which will eventually need to be pulled in for this PR.

camlow325 commented:
Rebased and added fec8582, which builds upon changes in the current ezbake PR at https://github.com/puppetlabs/ezbake/pull/410.  This will still fail Travis due to the jruby-utils and, now, ezbake SNAPSHOT dependencies.

camlow325 commented:
I noticed when running "puppetserver irb" with the previous code that I would get the following error:

```
Exception in thread "main" java.lang.NullPointerException: "in" is null!
	at java.nio.channels.Channels.checkNotNull(Channels.java:67)
	at java.nio.channels.Channels.newChannel(Channels.java:347)
	at org.jruby.util.io.ChannelDescriptor.open(ChannelDescriptor.java:783)
	at org.jruby.RubyFile.sysopen(RubyFile.java:1327)
	at org.jruby.RubyFile.sysopenInternal19(RubyFile.java:1307)
	at org.jruby.RubyFile.fileOpenGeneric(RubyFile.java:1240)
	at org.jruby.RubyFile.openFile19(RubyFile.java:1216)
	at org.jruby.RubyFile.initialize19(RubyFile.java:370)
	at org.jruby.RubyFile$INVOKER$i$0$2$initialize19.call(RubyFile$INVOKER$i$0$2$initialize19.gen)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:286)
	at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:81)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:85)
	at org.jruby.RubyClass.newInstance(RubyClass.java:879)
	at org.jruby.RubyIO.open(RubyIO.java:1197)
	at org.jruby.RubyIO$INVOKER$s$0$0$open.call(RubyIO$INVOKER$s$0$0$open.gen)
        ...
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_ROOT(ASTInterpreter.java:121)
	at org.jruby.Ruby.runInterpreter(Ruby.java:884)
	at org.jruby.Ruby.loadFile(Ruby.java:2837)
	at org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.loadScript(LibrarySearcher.java:264)
	at org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.load(LibrarySearcher.java:252)
	at org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:51)
	at org.jruby.runtime.load.LoadService.load(LoadService.java:358)
	at org.jruby.RubyKernel.loadCommon(RubyKernel.java:1067)
	at org.jruby.RubyKernel.load19(RubyKernel.java:1059)
	at org.jruby.RubyKernel$INVOKER$s$0$1$load19.call(RubyKernel$INVOKER$s$0$1$load19.gen)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:210)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:206)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:326)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:170)
	at org.jruby.ast.FCallOneArgNode.interpret(FCallOneArgNode.java:36)
	at org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
	at org.jruby.ast.RootNode.interpret(RootNode.java:129)
	at org.jruby.evaluator.ASTInterpreter.INTERPRET_ROOT(ASTInterpreter.java:121)
	at org.jruby.Ruby.runInterpreter(Ruby.java:884)
	at org.jruby.Ruby.runInterpreter(Ruby.java:892)
	at org.jruby.Ruby.runNormally(Ruby.java:733)
	at org.jruby.Ruby.runFromMain(Ruby.java:578)
	at org.jruby.Main.doRunFromMain(Main.java:393)
	at org.jruby.Main.internalRun(Main.java:288)
	at org.jruby.Main.run(Main.java:217)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:93)
	at clojure.lang.Reflector.invokeInstanceMethod(Reflector.java:28)
	at puppetlabs.services.jruby_pool_manager.jruby_core$eval8719$cli_ruby_BANG___8724$fn__8725.invoke(jruby_core.clj:305)
	at puppetlabs.services.jruby_pool_manager.jruby_core$eval8719$cli_ruby_BANG___8724.invoke(jruby_core.clj:299)
	at puppetlabs.services.jruby_pool_manager.jruby_core$eval8744$cli_run_BANG___8749$fn__8750.invoke(jruby_core.clj:316)
	at puppetlabs.services.jruby_pool_manager.jruby_core$eval8744$cli_run_BANG___8749.invoke(jruby_core.clj:307)
	at puppetlabs.puppetserver.cli.irb$irb_run_BANG_.invokeStatic(irb.clj:9)
	at puppetlabs.puppetserver.cli.irb$irb_run_BANG_.invoke(irb.clj:6)
	at puppetlabs.puppetserver.cli.subcommand$run.invokeStatic(subcommand.clj:44)
	at puppetlabs.puppetserver.cli.subcommand$run.invoke(subcommand.clj:33)
	at puppetlabs.puppetserver.cli.subcommand$jruby_run.invokeStatic(subcommand.clj:60)
	at puppetlabs.puppetserver.cli.subcommand$jruby_run.invoke(subcommand.clj:56)
	at puppetlabs.puppetserver.cli.irb$_main.invokeStatic(irb.clj:13)
	at puppetlabs.puppetserver.cli.irb$_main.doInvoke(irb.clj:11)
```

The culprit was a previous code change done for JRuby 9k compatibility, appending "classpath:/" to the default `ruby-code-dir`, used to add Puppet Server's embedded Ruby code to the JRuby load path.  When I removed the "classpath:/" for JRuby 1.7, the irb command started working again.  The behavior seemed similar to what was mentioned in https://github.com/jruby/jruby/issues/1983, although that specific problem had been addressed prior to the JRuby 1.7.26 version that we're using in Puppet Server currently.

Appending "classpath:/" works fine for JRuby 9k.  Since many further code changes for JRuby 1.7 are not likely at this point, I added a commit, 5033664, which allows the `ruby-code-dir` to be set without the "classpath:/" prefix for JRuby 1.7 (for backward compatibility) while including the prefix for later versions, since it is necessary for proper Ruby loading.  I made this change in 5033664.

camlow325 commented:
I edited the ezbake integration commit, now f46dd43, to work with the latest changes in the corresponding ezbake branch.  I'm running a build of that, http://builds.puppetlabs.lan/puppetserver/5.0.0.master.SNAPSHOT.2017.04.27T1516/, back through beaker tests both for JRuby 1.7 and 9k to confirm that tests are passing.

camlow325 commented:
I ran through each of the following sets of test suites under both JRuby 1.7 and 9k and all were passing.  For the beaker tests, I was running with the SNAPSHOT package I mentioned in a previous comment, for a single cell master/agent configuration on EL-7 only.

* Clojure unit tests
* Puppet Server Ruby spec tests
* Puppet Server smoke acceptance tests
* Ruby Puppet full acceptance tests

jpinsonault commented on test/integration/puppetlabs/services/jruby/jruby_puppet_pool_int_test.clj:
```diff
@@ -330,7 +330,12 @@
+               (testing "fixed at '2.1' if '2.1' version available"
+                 (is (= "RUBY2_1" (.. container (getCompatVersion) (toString)))))
+               (testing "returns '2.0' per compat version setting"
+                 (is (= CompatVersion/RUBY2_0 (.getCompatVersion container))))))
```
Might be nice for future readers to add the description from your commit about what this test is doing now

camlow325 replied:
I added a comment for this in a new commit, 23b4fcd.

jpinsonault commented on src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj:
```diff
@@ -31,7 +32,19 @@
+      ;; that the "classpath:/" prefix is only omitted for JRuby 1.x.
+      (str (if (not= "1." (subs jruby-version 0 2))
+             "classpath:/")
+           "puppetserver-lib"))))
```
Doing a substring match makes me mildly wary, but we're pretty confident about what the version string will contain I guess

camlow325 replied:
Yeah, I'm not thrilled about that either but I didn't see any other more straightforward way to interrogate JRuby to see if it is 1.7.x or 9k derivative.  If you see something else which seems better to you, happy to go with that instead.

jpinsonault replied:
Yeah I don't have any ideas other than regex, which, yeah...

camlow325 commented:
I added a couple of commits which use new functionality from jruby-utils for determining whether the current version being used on the server is JRuby 9k-based, made a number of test changes to work with the latest jruby-utils behavior for JRuby 9k, and removed the jruby-utils pin.  This PR should now pass Travis hopefully but still should not be merged until the ezbake version it has been pinned to has been released.  The ezbake release is in process.

rlinehan commented on project.clj:
```diff
@@ -36,14 +36,14 @@
   :min-lein-version "2.7.1"
 
-  :parent-project {:coords [puppetlabs/clj-parent "0.6.0"]
+  :parent-project {:coords [puppetlabs/clj-parent "0.7.0"]
```
I think 0.8.0 exists now?

camlow325 replied:
Good call, thanks.  I added a commit, 1c12109, which bumps the version.

rlinehan commented on resources/ext/cli/ruby.erb:
```diff
@@ -1,6 +1,18 @@
+    exit 1
+  fi
+fi
+
```
Is there any way we could abstract this somehow so its not copied betwen all three of these files? Or would that be more complicated than it's worth?

camlow325 replied:
We might be able to reduce the duplication a little bit.  Some of the duplication, though, seems desirable to me.  At least the bits about setting a default value for the CLASSPATH variable and checking if the cli_defaults file exists before trying to source it would seem good to duplicate between the files for error handling / robustness within the individual scripts.  My preference would be to leave this in for now but would be curious to hear if others, like @jpinsonault have any strong opinions about this.

rlinehan commented on resources/ext/build-scripts/install-vendored-gems.sh:
```diff
@@ -19,5 +19,5 @@ while read LINE
   gem_name=$(echo $LINE |awk '{print $1}')
   gem_version=$(echo $LINE |awk '{print $2}')
-  java -cp puppet-server-release.jar clojure.main -m puppetlabs.puppetserver.cli.gem --config jruby.conf -- install ${gem_name} --version ${gem_version}
+  java -cp puppet-server-release.jar:jruby-1_7.jar clojure.main -m puppetlabs.puppetserver.cli.gem --config jruby.conf -- install ${gem_name} --version ${gem_version}
```
this worries me a bit but also I guess it's probably not super bad because this is only installing a specific list of gems, so so long as we don't have any gems that don't work with 1.9.3 in there it's fine?

camlow325 replied:
Yeah, all of the gem versions we have pinned in there currently should work fine with Ruby language version 1.9.3.  Hopefully that's the case as well for JRuby 9k (currently Ruby language version 2.3.1) - at least I haven't seen any aberrant behavior from them when running Puppet acceptance tests under JRuby 9k so far.

At some point in time, it does seem like we could run into a situation where we need to have different sets of gems for 1.7 vs 9k.  At that point, we may need to modify this script to bundle gems into separate directories for each and modify the Puppet Server source to setup the appropriate vendor gem path by default for the current JRuby version in use.  I was hoping to avoid having to look into that until we may need to do it.

camlow325 commented:
ezbake 1.3.0 has been released so this should be ready to merge.  I did a build for EL-7 from the latest in this PR, http://builds.puppetlabs.lan/puppetserver/5.0.0.master.SNAPSHOT.2017.05.04T1131/.  I've run the Puppet Server smoke tests (including the new one which validates JRuby 9k) and all are passing.

rlinehan commented:
Alright, built some packages, played around on both centos 7 and ubuntu 16.04

Nice error when jruby jar is wrong.

```
Unable to find specified JRUBY_JAR: /opt/puppetlabs/server/apps/puppetserver/jruby-k.jar
```

Good job @jpinsonault and @camlow325 on that - super easy to understand!

I think it would be good to have an info-level log about what JRuby version is being used, maybe from Puppet Server itself, on startup. I just can imagine support cases where that's information we want pretty early on, but we have to tell users to change to debug and possibly restart to get that. I would like to merge this though and start working on getting it through CI, rather than waiting for that.

camlow325 commented:
> I think it would be good to have an info-level log about what JRuby version is being used, maybe from Puppet Server itself, on startup. I just can imagine support cases where that's information we want pretty early on, but we have to tell users to change to debug and possibly restart to get that.

Sorry, @rlinehan, took me a while to remember to get back to this. I think this is a great idea. I put up a separate PR, #1400, which adds a version log message. Hope this is something like what you had in mind.
